### PR TITLE
ENH: add 'whatrecord graph' entrypoint with at least db/st.cmd support

### PR DESCRIPTION
**Pull request by klauer**
## Description
Add command-line support for:
```
$ whatrecord graph st.cmd
$ whatrecord graph database.db
```

## Motivation and Context
It'd be nice to not have to spawn up a server to investigate record links.

## Screenshots (if appropriate):

Output of:
```
whatrecord graph pluginROI.db  --highlight "HOPR" |dot -Tpdf -O
```

image
(internal link fields used here; no dbd required)

### CODE REVIEW

**Comment by klauer** [comment]
Some forward-looking fixes for lark 1.0 (but pinning to < 1.0 for now) (see #113)
I think this is good enough for now; requesting rubber stamp/code smell check

File: whatrecord/bin/graph.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+"whatrecord graph" is used to parse and generate a relationship graph of
+whatrecord-supported file types.
+"""
+
+import argparse
+import logging
+import re
+from typing import Dict, List, Optional
+
+# from ..access_security import AccessSecurityConfig
```
Comment (ZLLentz) [inline whatrecord/bin/graph.py:11]: future plans, perhaps?

File: whatrecord/bin/graph.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+logger = logging.getLogger(__name__)
+DESCRIPTION = __doc__
+
+
+def build_arg_parser(parser=None):
+    if parser is None:
+        parser = argparse.ArgumentParser()
+
+    parser.description = DESCRIPTION
+    parser.formatter_class = argparse.RawTextHelpFormatter
```
Comment (ZLLentz) [inline whatrecord/bin/graph.py:33]: It doesn't seem like your help strings are pre-formatted. What does using the RawTextHelpFormatter help you do?

File: whatrecord/bin/graph.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        for filename in filenames
+    ]
+
+    if databases_only:
+        result = _combine_databases(*results)
+    else:
+        result, = results
+
+    if isinstance(result, (LinterResults, LoadedIoc, Database)):
+        if isinstance(result, LoadedIoc):
```
Comment (ZLLentz) [inline whatrecord/bin/graph.py:181]: I guess this kind of if/else tree construct is why they've added the match statement in 3.10

File: whatrecord/bin/parse.py [hunk 4]
```diff
@@ -144,6 +151,9 @@ def parse(
 
     expand : bool, optional
         Expand a substitutions file.
+
+    v3 : bool, optional
+        Use V3 database grammar where applicable.
```
Comment (ZLLentz) [inline whatrecord/bin/parse.py:156]: Is this referring to EPICS v3?

File: whatrecord/bin/graph.py (hunk 2 above)
Comment (klauer) [reply whatrecord/bin/graph.py:33]: Mostly helps with keeping `DESCRIPTION` in tact, if I ever get around to making it a bit more verbose. Good catch in that it's unnecessary here :+1: (it's a copy/paste job of all of the other subcommand entrypoints 😁 )

File: whatrecord/bin/graph.py (hunk 1 above)
Comment (klauer) [reply whatrecord/bin/graph.py:11]: Yeah, technically all could be used for graphing in _some way_ but I'm not sure what it'd look like.

I think at least state notation language would benefit from a state transition diagram one day.

File: whatrecord/bin/parse.py (hunk 4 above)
Comment (klauer) [reply whatrecord/bin/parse.py:156]: Yes, EPICS V3. Database grammar is a bit different from one to the other, believe it or not...

File: whatrecord/bin/graph.py (hunk 3 above)
Comment (klauer) [reply whatrecord/bin/graph.py:181]: I'd really like to be able to use it, shame we have to wait for 3.10...

**Review by ZLLentz — APPROVED** [review]
Some minor questions/observations that don't need to hold up the PR if you're trying to move on. This looks really useful.

File: whatrecord/bin/graph.py (hunk 1 above)
Comment (ZLLentz) [reply whatrecord/bin/graph.py:11]: Gotcha. It all stood out to me because I usually don't see tons of commented lines in your code.

**Comment by klauer** [comment]
Merging for now. Will revisit in the new year.

**Pull request merged**